### ISSUE TITLE
Use a weighted loss function instead of oversampling

### ISSUE THREAD

**Issue by lmbak** [azavea/raster-vision#867]
The oversampling functionality allows imbalanced classes to be shown equally during the training process. However, some form of randomness is included because images are randomly sampled. In an extreme case, or a case with few epochs, this would mean that not all images are used during training, which is undesirable as that would mean a loss of information shown to the model.

One way to solve this could be to use a weighted loss function. In the [Documentation](https://pytorch.org/docs/stable/nn.html) of the CrossEntropyLoss function it is explained that a weights vector can be passed to the loss function. This way the randomness would be removed from the training process, likely leading to more reproducible results.

**Comment by AdeelH** [comment]
I can think of multiple ways to accomplish this. In descending order of generality:

1. Add a `loss_fn` option to `SolverConfig`, set to [`nn.CrossEntropyLoss()`](https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html#torch.nn.CrossEntropyLoss) by default
	- the user will be able to pass in, for example, `nn.CrossEntropyLoss(weight=custom_class_weights)`
		- or something more complicated
2. Add a `class_loss_weights` option `SolverConfig` where the user can supply custom weights
3. Add a `weight_class_losses_equally` flag to `SolverConfig`, which will cause the class weights to be automatically calculated from the training set

@lewfish, @jamesmcclain, what do you guys think?

**Comment by lewfish** [comment]
> I can think of multiple ways to accomplish this. In descending order of generality:
> 
> 1. Add a `loss_fn` option to `SolverConfig`, set to [`nn.CrossEntropyLoss()`](https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html#torch.nn.CrossEntropyLoss) by default
>    
>    * the user will be able to pass in, for example, `nn.CrossEntropyLoss(weight=custom_class_weights)`
>      
>      * or something more complicated
>        - con: might not work well with config serialization
> 2. Add a `class_loss_weights` option `SolverConfig` where the user can supply custom weights
> 3. Add a `weight_class_losses_equally` flag to `SolverConfig`, which will cause the class weights to be automatically calculated from the training set
> 
> @lewfish, @jamesmcclain, what do you guys think?

I agree that 1 will be a problem for config serialization. 2 and 3 both seem like useful options.

**Comment by lewfish** [comment]
As discussed at standup, since you've added option 2 already, I would hold off on adding option 3 until we need it.
